### PR TITLE
schitems: parse symbol instances

### PR DESCRIPTION
**Pull request by hvraven**
This PR adds an implementation to parse the new symbol instances added in kicad 7.
This code works, but is not yet ready to merge as is. Parts of the documentation are missing and I did not check at all what happens to kicad 6 schematics.
I'll gladly polish it to a proper level, but wanted to get your input first whether you think the chosen data structure makes sense.

### CODE REVIEW

**Comment by eeintech** [comment]
Hello @hvraven 

I was looking into [this issue](https://github.com/mvnmgrx/kiutils/issues/70) and saw your PR which intends to fix the `SchematicSymbol` part of it (not the `HierarchicalSheet`).

I was wondering:
- Why not include a class variable for the project `name` inside your `SymbolProjectInstance` class? Then the `SchematicSymbol.instances` would be a list of `SymbolProjectInstance`  instead of a dictionary `{project: SymbolProjectInstance()}`
- In light of the fact that both `HierarchicalSheet` and `SchematicSymbol` will require it, how about naming it `ProjectInstance` instead to avoid naming confusion?
- Maybe consider moving `parse_project_instances` method content into the new `ProjectInstance.from_sexpr` method? 

Obviously @mvnmgrx would make the call ultimately, I am only proposing the changes I thought would make sense :smiley:

File: src/kiutils/items/schitems.py [hunk 1]
```diff
@@ -780,6 +780,10 @@ class SchematicSymbol():
     """The ``pins`` token defines a dictionary with pin numbers in form of strings as keys and
        uuid's as values"""
 
+    instances: Dict[str, List[SymbolProjectInstances]] = field(default_factory=dict)
```
Comment (mvnmgrx) [inline src/kiutils/items/schitems.py:783]: Please use a `List[ProjectInstanceNameToBeDefined]` here

File: src/kiutils/items/schitems.py [hunk 2]
```diff
@@ -816,9 +820,23 @@ def from_sexpr(cls, exp: list) -> SchematicSymbol:
             if item[0] == 'at': object.position = Position().from_sexpr(item)
             if item[0] == 'property': object.properties.append(Property().from_sexpr(item))
             if item[0] == 'pin': object.pins.update({item[1]: item[2][1]})
+            if item[0] == 'instances': object.instances = cls.parse_project_instances(item)
```
Comment (mvnmgrx) [inline src/kiutils/items/schitems.py:823]: Here we could loop over all tokens in `item[1]` to get to all project instances:
```python
if item[0] == 'instances':
  for instance in item[1]:
    object.instances.append(ProjectInstanceNameToBeDefined.from_sexpr(instance))
```

File: src/kiutils/items/schitems.py [hunk 3]
```diff
@@ -816,9 +820,23 @@ def from_sexpr(cls, exp: list) -> SchematicSymbol:
...
             if item[0] == 'mirror': object.mirror = item[1]
         return object
 
+    @staticmethod
+    def parse_project_instances(exp: list) -> Dict[str, SymbolProjectInstance]:
```
Comment (mvnmgrx) [inline src/kiutils/items/schitems.py:828]: This could be removed completely then

File: src/kiutils/items/schitems.py [hunk 4]
```diff
@@ -847,6 +865,71 @@ def to_sexpr(self, indent=2, newline=True) -> str:
             expression += property.to_sexpr(indent+2)
         for number, uuid in self.pins.items():
             expression += f'{indents}  (pin "{dequote(number)}" (uuid {uuid}))\n'
+        for project, instances in self.instances.items():
```
Comment (mvnmgrx) [inline src/kiutils/items/schitems.py:868]: Here we loop over just the items in `self.instances` then:
```python
if self.instances:
  expression += f'{indents}  (instances\n'
  for instance in self.instances:
    expression += instance.to_sexpr(indent+2)
  expression += f'{indents}  )\n'
```

File: src/kiutils/items/schitems.py [hunk 5]
```diff
@@ -847,6 +865,71 @@ def to_sexpr(self, indent=2, newline=True) -> str:
...
+            expression += f'{indents}    (project "{dequote(project)}"\n'
+            for instance in instances:
+                expression += instance.to_sexpr(indent+6)
+            expression += f'{indents}    )\n'
+            expression += f'{indents}  )\n'
+        expression += f'{indents}){endline}'
+        return expression
+
+@dataclass
+class SymbolProjectInstance:
```
Comment (mvnmgrx) [inline src/kiutils/items/schitems.py:879]: Lets discuss the naming scheme in the PR comments

**Comment by mvnmgrx** [comment]
I agree with all your points @eeintech. 

The only problem with making a common class for both `Symbol` and `HierarchicalSheet` is that they convey different meaning as they have different tokens inside. See here:
- [Hierarchical Sheet Section](https://dev-docs.kicad.org/en/file-formats/sexpr-schematic/#_hierarchical_sheet_section) (uses only a `page` token in its `path`)
- [Symbol Section](https://dev-docs.kicad.org/en/file-formats/sexpr-schematic/#_symbol_section) (uses `reference` and `unit` token in its `path`)

So we could either make some common class `ProjectInstance` that takes some list of `ProjectPath` elements that has all tokens inside and only populate the tokens we need. But i think that this approach is a bit ambiguous.

I would suggest that we add such a common class that has a `name: str` and a `paths: List[]` token. The `paths` token can then have elements of either `HierarchicalSheetProjectPath` as well as `SymbolProjectPath` that implement the neccessary tokens. All formatting and parsing can then be done in `to_sexpr()` and `from_sexpr()` as usual.

Example of this:
```python
@dataclass
class ProjectInstance():
  name: str
  paths: List[Union[HierarchicalSheetProjectPath, SymbolProjectPath]] = field(...)
  # ...

@dataclass
class HierarchicalSheetProjectPath():
  sheetInstancePath: str
  page: str
  # ...

@dataclass
class SymbolProjectPath():
  sheetInstancePath: str
  reference: str
  unit: int
  # ...
```

Or separate both into `SymbolProjectInstance` and `HierarchicalSheetProjectInstance` with the said subclasses in it to make it least ambiguous.

Happy to hear your thoughts on this.

**Review by mvnmgrx — CHANGES_REQUESTED** [review]
I've added some early comments, please check my other comment in the PR for further discussing.

**Comment by eeintech** [comment]
@mvnmgrx On-board with all your comments.

> Or separate both into `SymbolProjectInstance` and `HierarchicalSheetProjectInstance` with the said subclasses in it to make it least ambiguous.

I vote for the KISS philosophy, it makes more sense than what I proposed, I actually didn't realized they were different formats and variables to store.

@hvraven Do you think you can update your PR with the feedback? Happy to help out

**Comment by mvnmgrx** [comment]
> I vote for the KISS philosophy

I agree

> if SymbolProjectInstance and HierarchicalSheetProjectInstance classes happen to have a lot of methods in common then we could make them children of a "generic" ProjectInstance class maybe?

Currently, they only seem to share the `name` token. But this would enable us to add upcomming features more easily (at least i expect that they may add more project-related stuff in the future). 

This may look something like this:

```python
@dataclass
class ProjectInstance(ABC):
    name: str

    @abstractmethod
    def from_sexpr():
        raise NotImplementedError

    @abstractmethod
    def to_sexpr():
        raise NotImplementedError

@dataclass 
class SymbolProjectPath():
    ...

@dataclass
class SymbolProjectInstance(ProjectInstance):
    paths: List[SymbolProjectPath] = field(default_factory=list)
    # Implement from_sexpr and to_sexpr here

@dataclass
class HierarchicalSheetProjectPath():
    ...

@dataclass
class HierarchicalSheetProjectInstance(ProjectInstance):
    paths: List[HierarchicalSheetProjectPath] = field(default_factory=list)
    # Implement from_sexpr and to_sexpr here
```

I kind of like this approach as well. At least we would be more flexible if KiCad adds new features here. What do you think?

If we additionally add the `paths` token into the base class as well, then there would be nothing to inherit. It would look like the one i proposed above, so we have to make the two separate `xxxxPath` classes for the inheritance approach.

**Comment by mvnmgrx** [comment]
Okay, lets do it this way 👍 

Let me know if i should take over or if you need any help

**Comment by eeintech** [comment]
@mvnmgrx If you want to take over I think you would be faster than me, ATM I have very limited time on my hands but wish to release a new Ki-nTree version soon with a version of `kiutils` that supports KiCad V7 :smiley:

**Comment by mvnmgrx** [comment]
We should be mostly done with KiCad v7 stuff, if i haven't forgot anything they added. 

@hvraven do you want to implement the proposed changes in your PR or shall i take over?

**Comment by mvnmgrx** [comment]
This PR is the last major change before i want to release the next version that starts supporting KiCad v7. We then should have implemented every new feature that was added. But don't hesitate to open a issue if you find something thats not parsed correctly yet.

**Comment by hvraven** [comment]
Thanks for your work on that, the result looks good. I wasn't really sure how to best map the new properties into the class in the first place, the solution you two found looks a lot nicer then mine.
I will try to rework it today, if I do not come around to it and you find time feel free to take it instead.

**Comment by eeintech** [comment]
> This PR is the last major change before i want to release the next version that starts supporting KiCad v7. We then should have implemented every new feature that was added. But don't hesitate to open a issue if you find something thats not parsed correctly yet.

I tested adding a symbol into a KiCad V7 library using `1.3.0` release and got the [expected error](https://github.com/mvnmgrx/kiutils/issues/62) from stroke parsing:
``` shell
  File "/home/francois/.local/lib/python3.10/site-packages/kiutils/items/syitems.py", line 406, in from_sexpr
    if item[0] == 'stroke': object.stroke = Stroke().from_sexpr(item)
  File "/home/francois/.local/lib/python3.10/site-packages/kiutils/items/common.py", line 237, in from_sexpr
    raise Exception("Expression does not have the correct type")
Exception: Expression does not have the correct type
```

But with master, it works great and symbol passes our CI pipeline! (which is still based on [KiCad's kicad-library-utils](https://gitlab.com/kicad/libraries/kicad-library-utils) library)
Can't wait for this PR to be merged and the new release :smile:

**Pull request closed without merging**